ruby_programming/computer_science/project_data_structures_algorithms.md

```diff
@@ -15,34 +15,42 @@ You'll build a simple binary search tree in this assignment. In this lesson, our
+  9. Write a `#balanced?` method which checks if the tree is balanced. A balanced tree is one where the difference between heights of left subtree and right subtree is not more than 1.
 
-  9. Write a `#rebalance!` method which rebalances an unbalanced tree.
-  **Tip:** You'll want to create a level-order array of the tree before passing the array back into the `#build_tree` method.
+  10. Write a `#rebalance` method which rebalances an unbalanced tree. **Tip:** You'll want to create a level-order array of the tree before passing the array back into the `#build_tree` method.
 
-  10. Write a simple driver script that does the following:
+  11. Write a simple driver script that does the following:
 
     1. Create a binary search tree from an array of random numbers (`Array.new(15) { rand(1..100) }`)
     2. Confirm that the tree is balanced by calling `#balanced?`
     3. Print out all elements in level, pre, post, and in order
     4. try to unbalance the tree by adding several numbers > 100
     5. Confirm that the tree is unbalanced by calling `#balanced?`
-    6. Balance the tree by calling `#rebalance!`
+    6. Balance the tree by calling `#rebalance`
     7. Confirm that the tree is balanced by calling `#balanced?`
     8. Print out all elements in level, pre, post, and in order
 
-Pat yourself on the back! As a super-duper bonus, notice how all the depth-first methods share a similar signature and are basically just a re-arrangement of the same 3 lines... try dynamically declaring the three methods using metaprogamming techniques like `#define_method`.
+  **Tip:** If you would like to visualize your binary search tree, here is a `#pretty_print` method that a student wrote and shared on Discord:
+    ~~~ruby
+    def pretty_print(node = root, prefix="", is_left = true)
+      pretty_print(node.right, "#{prefix}#{is_left ? "│ " : " "}", false) if node.right 
+      puts "#{prefix}#{is_left ? "└── " : "┌── "}#{node.data.to_s}"
+      pretty_print(node.left, "#{prefix}#{is_left ? " " : "│ "}", true) if node.left
+    end
+    ~~~
+
+  ### Super-Duper Bonus
+  Did you notice how all the depth-first methods share a similar signature and are basically just a re-arrangement of the same 3 lines? If you would like to challenge yourself even further, here is an opportunity to learn about a metaprogamming technique called `#define_method`. This is not the best example of using metaprogramming, but it is important to know that it exists. You will have to do additional research to fully understand it, but this [Ruby Learning blog post on metaprogamming](http://rubylearning.com/blog/2010/11/23/dont-know-metaprogramming-in-ruby/), [Rubymonk's chapter on dynamic method calls](http://rubymonk.com/learning/books/2-metaprogramming-ruby/chapters/25-dynamic-methods/lessons/65-send#solution4186), and this [stack overflow](https://stackoverflow.com/questions/2306731/using-yield-inside-define-method-in-ruby) will get you started in the right direction.
```

Comment:
Writing these links as a list would be helpful. Users could quickly scan the links in a compact way.

Reply:
What do you think about removing this entire super-duper bonus? In this issue (https://github.com/TheOdinProject/curriculum/issues/17458) I mention Fensus and Timato talking about on Discord. They said things like "It's kinda a silly exercise though, should probably remove it" and "Yeah it's probably not the best use for pushing metaprogramming. And really doesn't show off it's power well."

However, another student heard about this being removed and suggested that it should stay in this issue (https://github.com/TheOdinProject/curriculum/issues/17956). 

I do not have enough Ruby experience to make an educated decision, so I would greatly appreciate your input before I make these changes. Thanks!

Reply:
Lets just remove it. If we want to introduce metaprogramming we shouldn't 1/2 ass a subject like that.